src/core/exec-invoke.c

```diff
@@ -2234,6 +2234,38 @@ static int setup_private_users(uid_t ouid, gid_t ogid, uid_t uid, gid_t gid) {
         return 0;
 }
 
+static int setup_private_pids(const ExecContext *c, ExecParameters *p) {
+        _cleanup_(pidref_done) PidRef pidref = PIDREF_NULL;
+        int r, q;
+
+        if (unshare(CLONE_NEWPID) < 0)
+                return log_exec_debug_errno(c, p, errno, "Failed to unshare pid namespace: %m");
+
+        /* The first process created after unsharing a pid namespace becomes PID 1 in the pid namespace, so
+         * we have to fork after unsharing the pid namespace to become PID 1. The parent sends the child
+         * pidref to the manager and exits while the child process continues with the rest of exec_invoke()
+         * and finally executes the actual payload. */
+
+        r = pidref_safe_fork("(sd-pidfd-child)", /*flags=*/ 0, &pidref);
+        if (r < 0)
+                return log_exec_debug_errno(c, p, r, "Failed to fork child into new pid namespace: %m");
+
+        if (r > 0) {
+                /* In the parent process, we send the child pidref to the manager and exit. */
```

Comment:
We usually do not free objects allocated by the parent process in the child process.
Is it safe to continue the remaining tasks in the child process??
Especially, the parent will free or close objects soon after the fork.

Reply:
@yuwata nspawn does the same thing where outer_child() sends some stuff to the parent and then immediately exits. I don't see how this is different from that.

Reply:
so I checked what exec_invoke() does after this, and I think we are probably fine (i.e. i see no NSS, no stdio and so on).

But this definitely deserves a comment, as requested.